### PR TITLE
feat: Slack Socket Mode for local development

### PR DESCRIPTION
**Pull request by amikofalvy**
## Summary

Add Slack Socket Mode support for local development so developers can receive Slack events over an outbound WebSocket — no ngrok, no tunneling, no dropped events on restart.

## Motivation

Local Slack development currently requires a tunneling tool (ngrok, Cloudflare Tunnel) to receive webhook events. This causes friction: tunnel setup, tunnel restarts on server restart, dropped events from disconnects, and latency. Slack Socket Mode eliminates all of this — the app initiates an outbound WebSocket to Slack, so events arrive directly to `localhost` with zero infrastructure.

## Approach

The core design extracts a **transport-agnostic event dispatcher** from the existing HTTP event handler. Both transports (HTTP webhooks in production, WebSocket in local dev) feed into the same `dispatchSlackEvent()` function, which routes events to the existing business logic handlers. This means local dev gets identical behavior and tracing to production.

```
Production:  Slack → HTTPS POST → sig verify → parse → dispatchSlackEvent() → handlers
Local dev:   Slack → WebSocket  → ack()      →         dispatchSlackEvent() → handlers
```

Socket Mode is opt-in and **development-only**: it activates when `ENVIRONMENT=development` and `SLACK_APP_TOKEN` is set. In production, Socket Mode is unconditionally disabled regardless of whether `SLACK_APP_TOKEN` is present — following the same `ENVIRONMENT === 'development'` guard used by other dev-only features. The setup script also warns if `ENVIRONMENT` is not set to `development`.

Each developer needs their own Slack app (Slack distributes Socket Mode events randomly across connections, not broadcast). A fully automated setup script (`pnpm setup-slack-dev`) creates the app via Slack API, captures OAuth tokens automatically, and configures `.env` — requiring only 2 manual pastes on first run and 0 on re-runs.

## Architectural decisions

| Decision | Choice | Rationale |
|---|---|---|
| Shared dispatcher vs. duplicated logic | Shared `dispatchSlackEvent()` | Single source of truth for event routing; both transports tested by same unit tests |
| `@slack/socket-mode` as devDependency | devDependency, dynamically imported | Zero production bundle impact; `MODULE_NOT_FOUND` caught gracefully if not installed |
| HMR safety | `globalThis` singleton | Prevents duplicate WebSocket connections during Vite HMR reloads; same pattern used by workflow world worker |
| One app per developer | Required by Slack platform | Socket Mode events are randomly distributed across connections (not broadcast), so shared apps would lose events |
| No `SLACK_SOCKET_MODE` flag | Infer from `SLACK_APP_TOKEN` presence | `SLACK_APP_TOKEN` is exclusively used for Socket Mode, so its presence is sufficient. Eliminates a redundant env var |
| Development-only guard | `ENVIRONMENT === 'development'` | Prevents accidental Socket Mode startup in production; same pattern as `createApp.ts`, `evalsAuth`, `workAppsAuth` |
| API-driven setup script | `apps.manifest.create` + auto-OAuth | Minimizes manual steps: 2 pastes first run (config refresh token + app-level token), 0 on re-runs |
| Dispatcher takes span parameter | Minimal interface `{ setAttribute, updateName }` | Avoids coupling dispatcher to `@opentelemetry/api`; both transports pass their own span objects |

### New files
- **`packages/agents-work-apps/src/slack/dispatcher.ts`** — Transport-agnostic event dispatcher extracted from `events.ts`. Handles `event_callback`, `block_actions`, `interactive_message`, `message_action`, `view_submission` routing. Returns `SlackEventDispatchResult` with optional `response` for view_submission validation errors.
- **`packages/agents-work-apps/src/slack/socket-mode.ts`** — Socket Mode adapter. Registers listeners for `slack_event`, `interactive`, and `slash_commands` events. Full OTel tracing with `flushTraces()` on every handler. `globalThis` singleton prevents duplicate connections on HMR.
- **`scripts/setup-slack-dev.ts`** — Fully automated setup script. Uses Slack Manifest API (`apps.manifest.create`) to create the dev app, captures OAuth bot token via a temporary localhost HTTP server, and writes all credentials to `.env`. Generates random two-noun dev identifiers (e.g., `frost-ember`) for collision-free multi-developer workspaces. State persisted in `.slack-dev.json` for idempotent re-runs. Colored output respects `NO_COLOR` env var. Warns if `ENVIRONMENT` is not set to `development`.
- **`packages/agents-work-apps/src/__tests__/slack/dispatcher.test.ts`** — 13 unit tests covering all event types and edge cases
- **`packages/agents-work-apps/src/__tests__/slack/socket-mode.test.ts`** — 11 unit tests covering all listener types and HMR safety

### Modified files
- **`packages/agents-work-apps/src/slack/routes/events.ts`** — Refactored to call `dispatchSlackEvent()` for event routing. Still owns HTTP-specific concerns: signature verification, retry dedup, body parsing, `url_verification`.
- **`packages/agents-work-apps/src/slack/index.ts`** — Added barrel exports for `dispatchSlackEvent`, `SlackEventDispatchResult`, `DispatchOptions`, `startSocketMode`.
- **`packages/agents-work-apps/src/env.ts`** — Added `SLACK_APP_TOKEN` (optional string for Socket Mode app-level token).
- **`agents-api/src/env.ts`** — Added `SLACK_APP_TOKEN` (optional string).
- **`agents-api/src/index.ts`** — Conditional Socket Mode startup: when `ENVIRONMENT === 'development'` and `SLACK_APP_TOKEN` is set, dynamically imports and starts Socket Mode with a 3s delay (same pattern as workflow world worker).
- **`packages/agents-work-apps/package.json`** — Added `@slack/socket-mode@^2.0.5` as devDependency.
- **`packages/agents-work-apps/src/slack/README.md`** — Added comprehensive "Local Development with Socket Mode" documentation section.
- **`.env.example`** — Added `SLACK_APP_TOKEN` under new "Slack Socket Mode" section.
- **`.gitignore`** — Added `.slack-dev.json` (contains dev app tokens, must not be committed).
- **`package.json`** — Added `setup-slack-dev` script.

## How to verify

1. **Without Socket Mode (no behavior change):** Start `pnpm dev` without `SLACK_APP_TOKEN` — server starts normally, no Socket Mode logs.
2. **With Socket Mode:** Set `SLACK_APP_TOKEN=xapp-...` and `ENVIRONMENT=development` in `.env`, run `pnpm dev` — look for "Slack Socket Mode client started" log after ~3s.
3. **Production safety:** Set `ENVIRONMENT=production` with `SLACK_APP_TOKEN` present — Socket Mode should NOT start.
4. **Setup script:** Run `pnpm setup-slack-dev` — the script creates the app via API, opens browser for OAuth, and writes all env vars automatically. Warns if `ENVIRONMENT` is not `development`.

### Automated verification

- [x] **Package exports:** dispatchSlackEvent, SlackEventDispatchResult, startSocketMode all exported from barrel
- [x] **Env schema:** SLACK_APP_TOKEN in both env.ts files; no SLACK_SOCKET_MODE flag
- [x] **Conditional startup:** agents-api/src/index.ts checks `ENVIRONMENT === 'development'` AND `SLACK_APP_TOKEN`, dynamic import, MODULE_NOT_FOUND handling
- [x] **HMR safety:** globalThis singleton key, existing client check, early return on HMR reload, client set after creation
- [x] **OTel tracing:** tracer, SLACK_SPAN_NAMES, SLACK_SPAN_KEYS used; flushTraces() called in all 3 handlers
- [x] **Transport agnosticism:** Dispatcher has no Hono/HTTP-specific code; events route still handles sig verify, body parsing, url_verification
- [x] **Ack patterns:** slack_event acks immediately before dispatch; interactive acks after dispatch (for view_submission response); slash_commands acks with command response
- [x] **Unit tests:** 162 pass across all Slack test files (13 dispatcher + 11 socket-mode + others)

### Manual verification required

- [x] **Real Slack connection:** Verify `pnpm dev` starts Socket Mode when `SLACK_APP_TOKEN` is set, @mention events received and traced
- [ ] **Slash commands via Socket Mode:** Verify slash commands work through WebSocket connection
- [x] **Interactive components:** Verify modals/button actions work via Socket Mode
- [x] **Without SLACK_APP_TOKEN:** Behavior unchanged — conditional startup only triggers when both conditions are met
- [x] **Production guard:** Socket Mode does not start when `ENVIRONMENT !== 'development'`
- [x] **Setup script flow:** `pnpm setup-slack-dev` creates app via API, captures OAuth token, writes env vars, warns if ENVIRONMENT is wrong
- [x] **Typecheck and lint pass**

## Future considerations

- **Signing secret verification for Socket Mode:** Socket Mode events are pre-authenticated by the WebSocket connection itself. No HMAC verification needed (or possible), but if Slack adds a signing mechanism for Socket Mode in the future, it should be added to the adapter.
- **Multi-workspace local dev:** Current design supports one workspace per dev app. If developers need to test across multiple workspaces simultaneously, the adapter would need to support multiple SocketModeClient instances.

---

Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: agents-api/src/index.ts [hunk 1]
```diff
@@ -129,4 +129,26 @@ if (workflowWorld === '@workflow/world-postgres' || workflowWorld === 'local') {
...
 
+// Start Slack Socket Mode client for local development (when configured)
+if (env.SLACK_SOCKET_MODE && env.SLACK_APP_TOKEN) {
+  const SOCKET_MODE_DELAY_MS = 3000;
+  console.log(`Scheduling Slack Socket Mode start in ${SOCKET_MODE_DELAY_MS}ms...`);
+
+  setTimeout(async () => {
+    try {
+      const { startSocketMode } = await import('@inkeep/agents-work-apps/slack');
+      await startSocketMode(env.SLACK_APP_TOKEN as string);
```
[the remark this answers is not in the document]
Comment (amikofalvy) [reply agents-api/src/index.ts]: Fixed in commit 3cee96e — added `SLACK_APP_TOKEN` and `SLACK_SOCKET_MODE` to `agents-api/src/env.ts` envSchema. The agents-api index.ts now correctly reads from the typed env object.

File: packages/agents-work-apps/src/slack/socket-mode.ts [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+    };
+
+    const response = await handleCommand(commandPayload);
+    await ack(
+      Object.keys(response).length > 0
+        ? (response as unknown as Record<string, unknown>)
+        : undefined
+    );
+    await flushTraces();
+  });
```
[the remark this answers is not in the document]
Comment (amikofalvy) [reply packages/agents-work-apps/src/slack/socket-mode.ts:168]: Fixed in commit cfa5746 — wrapped `slash_commands` handler in `tracer.startActiveSpan()` with `EVENT_TYPE`, `slack.command`, and `TEAM_ID` attributes. Error handling matches the pattern used by `slack_event` and `interactive` handlers.

File: packages/agents-work-apps/src/slack/socket-mode.ts [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+
+    await tracer.startActiveSpan(SLACK_SPAN_NAMES.WEBHOOK, async (span) => {
+      try {
+        const eventType = (body.event as { type?: string })?.type
+          ? 'event_callback'
+          : (body.type as string) || '';
+        span.setAttribute(SLACK_SPAN_KEYS.EVENT_TYPE, eventType);
+        span.updateName(`${SLACK_SPAN_NAMES.WEBHOOK} ${eventType}`);
+
+        await dispatchSlackEvent(eventType, body, { registerBackgroundWork: () => {} }, span);
```
[the remark this answers is not in the document]
Comment (amikofalvy) [reply packages/agents-work-apps/src/slack/socket-mode.ts]: Fixed in commit cfa5746 — replaced no-op `registerBackgroundWork` with a shared function that catches and logs errors. In Socket Mode (long-lived process), background work naturally completes since the process stays alive. The key improvement is error visibility — failed background work is now logged rather than silently swallowed.

File: packages/agents-work-apps/src/slack/socket-mode.ts [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+    return;
+  }
+
+  const { SocketModeClient } = await import('@slack/socket-mode');
+  const client = new SocketModeClient({ appToken });
+
+  setupSocketModeListeners(client);
+
+  (globalThis as Record<string, unknown>)[GLOBAL_KEY] = client;
+  await client.start();
```
[the remark this answers is not in the document]
Comment (amikofalvy) [reply packages/agents-work-apps/src/slack/socket-mode.ts]: Fixed in commit cfa5746 — added `error`, `disconnected`, and `reconnecting` event listeners to the SocketModeClient. Connection lifecycle events are now logged at appropriate levels (error/warn/info).

File: packages/agents-work-apps/src/slack/socket-mode.ts [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+    logger.info({}, 'Socket Mode client already running (HMR reload detected), skipping');
+    return;
+  }
+
+  const { SocketModeClient } = await import('@slack/socket-mode');
+  const client = new SocketModeClient({ appToken });
+
+  setupSocketModeListeners(client);
+
+  (globalThis as Record<string, unknown>)[GLOBAL_KEY] = client;
```
[the remark this answers is not in the document]
Comment (amikofalvy) [reply packages/agents-work-apps/src/slack/socket-mode.ts]: Fixed in commit cfa5746 — moved `globalThis` assignment to after `client.start()` succeeds. Added a test that verifies the timing: `globalThis` is undefined during `start()` and set only after it resolves.

File: packages/agents-work-apps/src/slack/socket-mode.ts [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+        const eventType = (body.event as { type?: string })?.type
+          ? 'event_callback'
+          : (body.type as string) || '';
+        span.setAttribute(SLACK_SPAN_KEYS.EVENT_TYPE, eventType);
+        span.updateName(`${SLACK_SPAN_NAMES.WEBHOOK} ${eventType}`);
+
+        await dispatchSlackEvent(eventType, body, { registerBackgroundWork: () => {} }, span);
+      } catch (error) {
+        span.setAttribute(SLACK_SPAN_KEYS.OUTCOME, 'error');
+        logger.error({ error }, 'Error handling Socket Mode event');
```
[the remark this answers is not in the document]
Comment (amikofalvy) [reply packages/agents-work-apps/src/slack/socket-mode.ts:77]: Fixed in commit cfa5746 — all three catch blocks now call `span.setStatus({ code: SpanStatusCode.ERROR })` and `span.recordException(error)`, matching the pattern in `events.ts`. `SpanStatusCode` is imported from `@opentelemetry/api` (already a dependency of this package).

**Comment by amikofalvy** [comment]
## Addressed "Consider" items from review

All 4 optional improvements from the review have been addressed in commit b2705cb:

### 1. Silent return when workspace lacks botToken (`dispatcher.ts:189-191`)
Added `logger.warn({ teamId }, 'Workspace found but no botToken — skipping modal project update')` so workspace configuration issues are visible in logs.

### 2. Inconsistent error notification for follow-up modal (`dispatcher.ts:260-266`)
Added `sendResponseUrlMessage` error notification in the `open_follow_up_modal` catch block, consistent with the `open_agent_selector_modal` pattern. Users now see an ephemeral "something went wrong" message when the follow-up modal fails to open.

### 3. No error handling for manifest JSON.parse (`setup-slack-dev.ts:24-26`)
Wrapped `JSON.parse` in try-catch with a contextual error message that includes the manifest file path.

### 4. Test coverage gaps for error handling paths
Added 4 new tests:
- `dispatchSlackEvent` errors in `slack_event` listener (verifies ack still called)
- `dispatchSlackEvent` errors in `interactive` listener (verifies ack called with undefined)
- `handleCommand` errors in `slash_commands` listener (verifies empty ack on failure)
- `handleAppMention` rejection (verifies background work still registered)

All 24 tests passing (13 dispatcher + 11 socket-mode). Typecheck and lint green.
